File: Cisco/4DecodeWays.py
from typing import Optional


class Solution:
    def numDecodings(self, s: str) -> int:
        N = len(s)

        def is_valid(a: str, b: Optional[str] = None) -> bool:
            if b:
                return a == "1" or (a == "2" and b < "7")
            return a != "0"

        prevprev = 1
        prev = 1 if is_valid(s[-1]) else 0

        for i in reversed(range(N - 1)):
            curr = 0
            if is_valid(s[i]):
                curr += prev
            if is_valid(s[i], s[i + 1]):
                curr += prevprev

            prev, prevprev = curr, prev

        return prev

File: Cisco/GitHubList/C++/4DecodeWays.cpp
#include <string>
#include <vector>

using namespace std;

class Solution
{
public:
    int numDecodings(string s)
    {
        int N = s.size();
        const auto &isValid = [&](const char &curr, const char &prev = 0)
        {
            if (prev != 0)
                return curr == '1' or (curr == '2' and prev >= '0' and prev <= '6');
            return curr != '0';
        };

        int p = s[N - 1] != '0' ? 1 : 0, pp = 1;
        for (int i = N - 2; i >= 0; --i)
        {
            int c = 0;
            if (isValid(s[i]))
                c += p;
            if (isValid(s[i], s[i + 1]))
                c += pp;

            pp = p, p = c;
        }

        return p;
    }
};